consul_exporter.go

```diff
@@ -0,0 +1,216 @@
+package main
+
+import (
+	"flag"
+	_ "fmt"
+	_ "io"
+	_ "io/ioutil"
+	"log"
+	_ "net"
+	"net/http"
+	_ "strconv"
+	_ "strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	namespace = "consul"
+)
+
+var (
+	serviceLabelNames = []string{"service", "node"}
+	memberLabelNames  = []string{"member"}
+)
+
+// Exporter collects HAProxy stats from the given URI and exports them using
+// the prometheus metrics package.
+type Exporter struct {
+	URI   string
+	mutex sync.RWMutex
+
+	up, clusterServers              prometheus.Gauge
+	totalQueries, jsonParseFailures prometheus.Counter
+	serviceMetrics, lockMetrics     map[string]*prometheus.GaugeVec
+	client                          *api.Client
+}
+
+func newServiceMetric(metricName string, docString string, constLabels prometheus.Labels) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace:   namespace,
+			Name:        "service_" + metricName,
+			Help:        docString,
+			ConstLabels: constLabels,
+		},
+		serviceLabelNames,
+	)
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(uri string, consulLocks string, timeout time.Duration) *Exporter {
+	// connect to Consul
+
+	consul_client, _ := api.NewClient(&api.Config{
+		Address: uri,
+	})
+
+	// init our exporter
+
+	return &Exporter{
+		URI: uri,
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "up",
+			Help:      "Was the last query of Consul successful.",
+		}),
+		totalQueries: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_total_queries",
+			Help:      "Current total Consul queries.",
+		}),
+
+		clusterServers: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "cluster_servers",
+			Help:      "How many peers are in the cluster.",
+		}),
+
+		serviceMetrics: map[string]*prometheus.GaugeVec{
+			"nodes_healthy":   newServiceMetric("nodes", "Number of nodes", prometheus.Labels{"healthy": "healthy"}),
```

Comment:
Hmm, this is a boolean. What you want then is a single metric called `node_healthy` with (service, node) as labels and a value of 0 or 1. That should let you do everything that you'd want to do with that data without getting into complex label matching.